p2panda-rs/src/operation/validate.rs

```diff
@@ -0,0 +1,338 @@
+/// 6. Check if payload matches claimed hash and size (#E6)
+/// 7. Correct hexadecimal operation encoding (when using human-readable encoding format) (#OP1)
+/// 8. Correct operation format as per specification, including canonic format checks against
+///    duplicate and unsorted operation fields (#OP2)
+/// 9. Correctly formatted and canonic operation field values, like document view ids (no
+///    duplicates, sorted, when no semantic value is given by that) as per specification (#OP3)
+/// 10. Operation fields match the claimed schema (#OP4)
+///
+/// ```text
+///                                                                  Look-Up
+///
+///             ┌────────────┐                       ┌─────┐    ┌─────┐    ┌─────┐
+///  bytes ───► │EncodedEntry├────decode_entry()────►│Entry│    │Entry│    │Entry│
+///             └──────┬─────┘                       └──┬──┘    └─────┘    └─────┘
+///                    │                                │
+///                    └───────────────────────────┐    │       Skiplink   Backlink
+///                                                │    │          │          │
+///             ┌────────────────┐                 │    │          │          │
+///  bytes ───► │EncodedOperation├─────────────┐   │    │          │          │
+///             └───────┬────────┘             │   │    │          │          │
+///                     │                      │   │    │          │          │
+///             decode_operation()             │   │    │          │          │
+///                     │            Look-Up   │   │    │          │          │
+///                     ▼                      │   │    │          │          │
+///              ┌──────────────┐    ┌──────┐  │   │    │          │          │
+///              │PlainOperation│    │Schema│  │   │    │          │          │
+///              └──────┬───────┘    └──┬───┘  │   │    │          │          │
+///                     │               │      │   │    │          │          │
+///                     │               │      │   │    │          │          │
+///                     │               │      │   │    │          │          │
+///                     │               │      │   │    │          │          │
+///                     │               ▼      ▼   ▼    ▼          ▼          │
+///                     └───────────►  validate_operation_and_entry() ◄───────┘
+///                                                 │
+///                                                 │
+///                                                 │
+///                                                 │
+///                                                 ▼
+///                                         ┌─────────────────┐
+///                                         │VerifiedOperation│
+///                                         └─────────────────┘
+/// ```
+#[allow(clippy::too_many_arguments)]
+pub fn validate_operation_with_entry(
```

Comment:
Things are sinking into my brain a little more and now I wonder about a more general point this threw up for me. It's a bit of an update on the previous comment, but I'll leave that there for posterity in any case.

Maybe the API for creating operations is a little too tied to `Schema` now. Obviously I understand the place this is coming from (making guarantees about the validity of data on it's creation), but I it could be ok to create operations optimistically (without their schema) as well. Of course we need methods for actually verifying them against their stated schema, but maybe this doesn't need to be enforced at this level. Operations which actually get _published_ will for sure be rejected by a node if they don't match their schema, but on the client side, maybe nice to just chuck out operations while only knowing the schema id. 

I could imagine 3 different useful APIs:

- create operations (not validated against a schema) with `Operation` or an `OperationBuilder`
- validate these operations against a schema with `validate_operation(&operation, &schema)`
- create and validate an operation with `Schema::create()`, `::update()` & `::delete()` (sure there are a million variations on this, but here's one idea)

Splitting up the operation creation into validated/un-validated would allow for a flow where we optimistically create operations without the schema, and the one we have, where we validate at the time of creation. I _think_ both these flows are valuable.

Reply:
If these flows are split up, then `VerifiedOperation` could become the one which has all guarantees, and `Operation` could still be useful (you can publish it) but not with quite the same guarantees as `VerifiedOperation`.

Reply:
related to this: https://github.com/p2panda/p2panda/pull/415#issuecomment-1209490662 which approaches the same issue by making `Schema` very easy to create :+1:

Reply:
This conversation reminds me of front-end data validation and back-end data validation, like, validating data in a form before sending it to the server is _better_ but you don't have to :rofl: it'll get rejected anyway.

Reply:
> I could imagine 3 different useful APIs:

Yeah, thats also a good idea, I'd be up for going with 1. and 2.! 3. is cool, but could be done later :+1:

Reply:
Yeh, I was imagining 3 as a "future" thing too.

Reply:
Do you want me to implement this for you to check after?

Reply:
> Do you want me to implement this for you to check after?

Yes, that's great! We can check with @cafca maybe first to get consensus, but I'm for it ✌️

Reply:
I made a very light touch implementation in this PR: https://github.com/p2panda/p2panda/pull/427

Reply:
@cafca do you have any impression on this change (I know it may be hard without looking at _everything_)?

The key takeaway I think, is that you can now construct an `Operation` without a `Schema` and without validating the fields. All the original methods for "validation max" remain where everything _is_ validated. We just opened a back door for easy operation creation.

Reply:
@adzialocha an alternate approach to what I implemented would be to offer an API for constructing `PlainOperation`s. This would then leave the `PlainOperation` -> `Operation` -> `VerifiedOperation` hierarchy un-touched and still offer a way to create/publish operations without needing the schema.

Reply:
That's what you did in `wasm` world... maybe that's nicer....

Reply:
An important case to consider in this convo is operations coming from the db in `aquadoggo`. We just want to instantiate an operation in this case, without the schema validation, as we trust the data.

Reply:
> @adzialocha an alternate approach to what I implemented would be to offer an API for constructing `PlainOperation`s. This would then leave the `PlainOperation` -> `Operation` -> `VerifiedOperation` hierarchy un-touched and still offer a way to create/publish operations without needing the schema.

I'm :+1: ing my own suggestion, think it's better. (partly cos then we don't need to change any of the diagrams :rofl: )

Reply:
> I think we still should go for Operation being without schemas.

Ok, coool